File: solutions/csharp/snakeandladdergame/SnakeAndLadderGame.cs
using System;
using System.Collections.Generic;

namespace SnakeAndLadderGame
{
    public class SnakeAndLadderGame
    {
        private readonly Board _board;
        private readonly List<Player> _players;
        private readonly Dice _dice;
        private int _currentPlayerIndex;

        public SnakeAndLadderGame(List<string> playerNames)
        {
            _board = new Board();
            _dice = new Dice();
            _players = new List<Player>();
            foreach (var playerName in playerNames)
            {
                _players.Add(new Player(playerName));
            }
            _currentPlayerIndex = 0;
        }

        public void Play()
        {
            while (!IsGameOver())
            {
                Player currentPlayer = _players[_currentPlayerIndex];
                int diceRoll = _dice.Roll();
                int newPosition = currentPlayer.Position + diceRoll;

                if (newPosition <= _board.GetBoardSize())
                {
                    currentPlayer.SetPosition(_board.GetNewPositionAfterSnakeOrLadder(newPosition));
                    Console.WriteLine($"{currentPlayer.Name} rolled a {diceRoll} and moved to position {currentPlayer.Position}");
                }

                if (currentPlayer.Position == _board.GetBoardSize())
                {
                    Console.WriteLine($"{currentPlayer.Name} wins!");
                    break;
                }

                _currentPlayerIndex = (_currentPlayerIndex + 1) % _players.Count;
            }
        }

        private bool IsGameOver()
        {
            foreach (var player in _players)
            {
                if (player.Position == _board.GetBoardSize())
                {
                    return true;
                }
            }
            return false;
        }
    }
}

File: solutions/python/snakeandladdergame/snake_and_ladder_game.py
from board import Board
from dice import Dice
from player import Player

class SnakeAndLadderGame:
    def __init__(self, player_names):
        self.board = Board()
        self.dice = Dice()
        self.players = [Player(name) for name in player_names]
        self.current_player_index = 0

    def play(self):
        while not self._is_game_over():
            current_player = self.players[self.current_player_index]
            dice_roll = self.dice.roll()
            new_position = current_player.get_position() + dice_roll

            if new_position <= self.board.get_board_size():
                current_player.set_position(self.board.get_new_position_after_snake_or_ladder(new_position))
                print(f"{current_player.get_name()} rolled a {dice_roll} and moved to position {current_player.get_position()}")

            if current_player.get_position() == self.board.get_board_size():
                print(f"{current_player.get_name()} wins!")
                break

            self.current_player_index = (self.current_player_index + 1) % len(self.players)

    def _is_game_over(self):
        for player in self.players:
            if player.get_position() == self.board.get_board_size():
                return True
        return False